sdk/keyvault/Azure.Security.KeyVault.Keys/src/Cryptography/LocalCryptographyClient.cs

```diff
@@ -0,0 +1,609 @@
+        /// <param name="plaintext">The data to encrypt.</param>
+        /// <param name="cancellationToken">A <see cref="CancellationToken"/> to cancel the operation.</param>
+        /// <returns>
+        /// The result of the encrypt operation. The returned <see cref="EncryptResult"/> contains the encrypted data
+        /// along with all other information needed to decrypt it. This information should be stored with the encrypted data.
+        /// </returns>
+        /// <exception cref="ArgumentException">The specified <paramref name="algorithm"/> does not match the key corresponding to the key identifier.</exception>
+        /// <exception cref="InvalidOperationException">The key is invalid for the current operation.</exception>
+        /// <exception cref="NotSupportedException">The operation is not supported with the specified key.</exception>
+        public virtual async Task<EncryptResult> EncryptAsync(EncryptionAlgorithm algorithm, byte[] plaintext, CancellationToken cancellationToken = default)
+        {
+            EncryptResult result = null;
+            if (_provider.SupportsOperation(KeyOperation.Encrypt))
+            {
+                result = await _provider.EncryptAsync(algorithm, plaintext, cancellationToken).ConfigureAwait(false);
+            }
+
+            return result ?? throw LocalCryptographyProvider.ThrowOperationNotSupported(nameof(KeyOperation.Encrypt));
+        }
+
+        /// <summary>
+        /// Encrypts the specified plain text.
+        /// </summary>
+        /// <param name="algorithm">The <see cref="EncryptionAlgorithm"/> to use.</param>
+        /// <param name="plaintext">The data to encrypt.</param>
+        /// <param name="cancellationToken">A <see cref="CancellationToken"/> to cancel the operation.</param>
+        /// <returns>
+        /// The result of the encrypt operation. The returned <see cref="EncryptResult"/> contains the encrypted data
+        /// along with all other information needed to decrypt it. This information should be stored with the encrypted data.
+        /// </returns>
+        /// <exception cref="ArgumentException">The specified <paramref name="algorithm"/> does not match the key corresponding to the key identifier.</exception>
+        /// <exception cref="InvalidOperationException">The key is invalid for the current operation.</exception>
+        /// <exception cref="NotSupportedException">The operation is not supported with the specified key.</exception>
+        public virtual EncryptResult Encrypt(EncryptionAlgorithm algorithm, byte[] plaintext, CancellationToken cancellationToken = default)
+        {
+            EncryptResult result = null;
+            if (_provider.SupportsOperation(KeyOperation.Encrypt))
+            {
+                result = _provider.Encrypt(algorithm, plaintext, cancellationToken);
+            }
+
+            return result ?? throw LocalCryptographyProvider.ThrowOperationNotSupported(nameof(KeyOperation.Encrypt));
```

Comment:
It looks like there are other reasons `result` could be null. Should we throw here only if `!_provider.SupportsOperation(KeyOperation.Encrypt)`?

Reply:
same for  many of the other methods below.

Reply:
All the ways `result` could be null still come down to not being supported. Cryptographic exceptions will still throw, but the original `LocalCryptographyProvider` as designed for local-or-remote operations returned `null` in graceful ways to facilitate falling through to the remote provider. So this logic still works out.

Was there any place in particular with which you were still concerned after my explanation?

Reply:
The only one I spot checked was [here](https://github.com/Azure/azure-sdk-for-net/blob/d00ac4a8f174e7605903c3e0bb80bc74b8f65f77/sdk/keyvault/Azure.Security.KeyVault.Keys/src/Cryptography/RsaCryptographyProvider.cs#L45-L55), but I didn't look deeply into all the ways the ciphertext could be null. Either way, it seems more strictly correct to only call `ThrowOperationNotSupported` for the explicit reason of `SupportsOperation` returning false.

Reply:
Thanks for the code pointer. Honestly, I'm not sure why I had that check - maybe a previous iteration of the code where it mattered. There's no practical scenario `RSA.Encrypt` should return null, so I could even remove the check (though, practically, it doesn't hurt). I'll go through other scenarios as well, but this should be fine. As discussed offline, the idea was to fall through to the remote provider without using exceptions as code flow. In practice, the only common scenario this may be null is that the customer provided a key without the private key material. If the JWK is junk (bad deserialization?), it will either not even get this far, or cause throw a `CryptographicException` later on - those still throw out of this call stack.

Reply:
In all cases I re-examined, return null means something of the algorithm is not supported by our client. Given the fall through to the remote provider, I do log ETW events with more specific details. We can consider conditionally throwing exceptions if doing local-only.

Reply:
https://github.com/Azure/azure-sdk-for-net/issues/14734